File: services/auth/src/main/java/com/example/iter/auth/service/RefreshTokenService.java
package com.example.iter.auth.service;

import com.example.iter.auth.domain.entity.RefreshToken;
import com.example.iter.auth.domain.entity.User;
import com.example.iter.common.security.UserStatus;
import com.example.iter.auth.domain.repository.RefreshTokenRepository;
import com.example.iter.auth.domain.repository.UserRepository;
import com.example.iter.auth.exception.RefreshTokenReuseException;
import com.example.iter.auth.service.model.IssuedTokenPair;
import com.example.iter.common.exception.CustomException;
import com.example.iter.common.exception.ErrorCode;
import com.example.iter.common.security.JwtTokenProvider;
import com.example.iter.common.security.TokenStatus;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class RefreshTokenService {

    private final RefreshTokenRepository refreshTokenRepository;
    private final UserRepository userRepository;
    private final RefreshTokenHasher refreshTokenHasher;
    private final JwtTokenProvider jwtTokenProvider;

    public String issueForLogin(User user) {
        if (user.getId() == null) {
            throw new IllegalArgumentException("저장되지 않은 회원에게 Refresh Token을 발급할 수 없습니다.");
        }

        String rawRefreshToken = jwtTokenProvider.generateRefreshToken(user.toAuthUser());
        RefreshToken refreshToken = RefreshToken.builder()
                .userId(user.getId())
                .tokenHash(refreshTokenHasher.hash(rawRefreshToken))
                .familyId(UUID.randomUUID().toString())
                .expiresAt(jwtTokenProvider.getExpiration(rawRefreshToken))
                .build();

        refreshTokenRepository.save(refreshToken);
        return rawRefreshToken;
    }

    @Transactional(noRollbackFor = RefreshTokenReuseException.class)
    public IssuedTokenPair rotate(String rawRefreshToken) {
        validateRefreshToken(rawRefreshToken);

        String tokenHash = refreshTokenHasher.hash(rawRefreshToken);
        RefreshToken currentToken = refreshTokenRepository.findWithLockByTokenHash(tokenHash)
                .orElseThrow(() -> new CustomException(ErrorCode.INVALID_REFRESH_TOKEN));

        if (!currentToken.getUserId().equals(jwtTokenProvider.getUserId(rawRefreshToken))) {
            throw new CustomException(ErrorCode.INVALID_REFRESH_TOKEN);
        }

        LocalDateTime now = LocalDateTime.now();
        if (currentToken.isRevoked()) {
            refreshTokenRepository.revokeAllActiveByFamilyId(currentToken.getFamilyId(), now);
            log.warn("Refresh Token 재사용 탐지: userId={}, familyId={}",
                    currentToken.getUserId(), currentToken.getFamilyId());
            throw new RefreshTokenReuseException();
        }
        if (currentToken.isExpired(now)) {
            throw new CustomException(ErrorCode.REFRESH_TOKEN_EXPIRED);
        }

        User user = userRepository.findById(currentToken.getUserId())
                .orElseThrow(() -> new CustomException(ErrorCode.INVALID_REFRESH_TOKEN));
        validateUserStatus(user);

        String newRawRefreshToken = jwtTokenProvider.generateRefreshToken(user.toAuthUser());
        RefreshToken replacementToken = RefreshToken.builder()
                .userId(user.getId())
                .tokenHash(refreshTokenHasher.hash(newRawRefreshToken))
                .familyId(currentToken.getFamilyId())
                .expiresAt(jwtTokenProvider.getExpiration(newRawRefreshToken))
                .deviceInfo(currentToken.getDeviceInfo())
                .build();

        RefreshToken savedReplacement = refreshTokenRepository.saveAndFlush(replacementToken);
        currentToken.rotateTo(savedReplacement.getId(), now);

        log.info("인증 토큰 재발급 처리: userId={}", user.getId());

        return new IssuedTokenPair(
                jwtTokenProvider.generateAccessToken(user.toAuthUser()),
                newRawRefreshToken
        );
    }

    @Transactional
    public void revoke(Long authenticatedUserId, String rawRefreshToken) {
        String tokenHash = refreshTokenHasher.hash(rawRefreshToken);
        refreshTokenRepository.findWithLockByTokenHash(tokenHash)
                .filter(token -> token.getUserId().equals(authenticatedUserId))
                .filter(token -> !token.isRevoked())
                .ifPresent(token -> token.revoke(LocalDateTime.now()));
    }

    @Transactional
    public void revokeAllByUserId(Long userId) {
        refreshTokenRepository.revokeAllActiveByUserId(userId, LocalDateTime.now());
    }

    private void validateRefreshToken(String rawRefreshToken) {
        TokenStatus status = jwtTokenProvider.validateRefreshToken(rawRefreshToken);
        if (status == TokenStatus.EXPIRED) {
            throw new CustomException(ErrorCode.REFRESH_TOKEN_EXPIRED);
        }
        if (status != TokenStatus.VALID) {
            throw new CustomException(ErrorCode.INVALID_REFRESH_TOKEN);
        }
    }

    private void validateUserStatus(User user) {
        if (user.getStatus() == UserStatus.SUSPENDED) {
            throw new CustomException(ErrorCode.USER_SUSPENDED);
        }
        if (user.getStatus() == UserStatus.DELETED) {
            throw new CustomException(ErrorCode.USER_DELETED);
        }
    }
}

File: services/auth/src/main/kotlin/com/example/iter/auth/service/RefreshTokenService.kt
package com.example.iter.auth.service

import com.example.iter.auth.domain.entity.RefreshToken
import com.example.iter.auth.domain.entity.User
import com.example.iter.auth.domain.repository.RefreshTokenRepository
import com.example.iter.auth.domain.repository.UserRepository
import com.example.iter.auth.exception.RefreshTokenReuseException
import com.example.iter.auth.service.model.IssuedTokenPair
import com.example.iter.common.exception.CustomException
import com.example.iter.common.exception.ErrorCode
import com.example.iter.common.security.JwtTokenProvider
import com.example.iter.common.security.TokenStatus
import com.example.iter.common.security.UserStatus
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.UUID

private val log = LoggerFactory.getLogger(RefreshTokenService::class.java)

@Service
class RefreshTokenService(
    private val refreshTokenRepository: RefreshTokenRepository,
    private val userRepository: UserRepository,
    private val refreshTokenHasher: RefreshTokenHasher,
    private val jwtTokenProvider: JwtTokenProvider,
) {

    fun issueForLogin(user: User): String {
        val userId = user.id
            ?: throw IllegalArgumentException("저장되지 않은 회원에게 Refresh Token을 발급할 수 없습니다.")

        val rawRefreshToken = jwtTokenProvider.generateRefreshToken(user.toAuthUser())
        val refreshToken = RefreshToken.builder()
            .userId(userId)
            .tokenHash(refreshTokenHasher.hash(rawRefreshToken))
            .familyId(UUID.randomUUID().toString())
            .expiresAt(jwtTokenProvider.getExpiration(rawRefreshToken))
            .build()

        refreshTokenRepository.save(refreshToken)
        return rawRefreshToken
    }

    @Transactional(noRollbackFor = [RefreshTokenReuseException::class])
    fun rotate(rawRefreshToken: String): IssuedTokenPair {
        validateRefreshToken(rawRefreshToken)

        val tokenHash = refreshTokenHasher.hash(rawRefreshToken)
        val currentToken = refreshTokenRepository.findWithLockByTokenHash(tokenHash)
            .orElseThrow { CustomException(ErrorCode.INVALID_REFRESH_TOKEN) }

        if (currentToken.userId != jwtTokenProvider.getUserId(rawRefreshToken)) {
            throw CustomException(ErrorCode.INVALID_REFRESH_TOKEN)
        }

        val now = LocalDateTime.now()
        if (currentToken.isRevoked()) {
            refreshTokenRepository.revokeAllActiveByFamilyId(currentToken.familyId, now)
            log.warn(
                "Refresh Token 재사용 탐지: userId={}, familyId={}",
                currentToken.userId,
                currentToken.familyId,
            )
            throw RefreshTokenReuseException()
        }
        if (currentToken.isExpired(now)) {
            throw CustomException(ErrorCode.REFRESH_TOKEN_EXPIRED)
        }

        val user = userRepository.findById(currentToken.userId)
            .orElseThrow { CustomException(ErrorCode.INVALID_REFRESH_TOKEN) }
        validateUserStatus(user)

        val newRawRefreshToken = jwtTokenProvider.generateRefreshToken(user.toAuthUser())
        val replacementToken = RefreshToken.builder()
            .userId(user.id!!)
            .tokenHash(refreshTokenHasher.hash(newRawRefreshToken))
            .familyId(currentToken.familyId)
            .expiresAt(jwtTokenProvider.getExpiration(newRawRefreshToken))
            .deviceInfo(currentToken.deviceInfo)
            .build()

        val savedReplacement = refreshTokenRepository.saveAndFlush(replacementToken)
        currentToken.rotateTo(savedReplacement.id, now)

        log.info("인증 토큰 재발급 처리: userId={}", user.id)

        return IssuedTokenPair(
            jwtTokenProvider.generateAccessToken(user.toAuthUser()),
            newRawRefreshToken,
        )
    }

    @Transactional
    fun revoke(authenticatedUserId: Long, rawRefreshToken: String) {
        val tokenHash = refreshTokenHasher.hash(rawRefreshToken)
        refreshTokenRepository.findWithLockByTokenHash(tokenHash)
            .filter { token -> token.userId == authenticatedUserId }
            .filter { token -> !token.isRevoked() }
            .ifPresent { token -> token.revoke(LocalDateTime.now()) }
    }

    @Transactional
    fun revokeAllByUserId(userId: Long) {
        refreshTokenRepository.revokeAllActiveByUserId(userId, LocalDateTime.now())
    }

    private fun validateRefreshToken(rawRefreshToken: String) {
        val status = jwtTokenProvider.validateRefreshToken(rawRefreshToken)
        if (status == TokenStatus.EXPIRED) {
            throw CustomException(ErrorCode.REFRESH_TOKEN_EXPIRED)
        }
        if (status != TokenStatus.VALID) {
            throw CustomException(ErrorCode.INVALID_REFRESH_TOKEN)
        }
    }

    private fun validateUserStatus(user: User) {
        if (user.status == UserStatus.SUSPENDED) {
            throw CustomException(ErrorCode.USER_SUSPENDED)
        }
        if (user.status == UserStatus.DELETED) {
            throw CustomException(ErrorCode.USER_DELETED)
        }
    }
}
